packages/gatsby/src/schema/infer/inference-metadata.js

```diff
@@ -63,16 +63,51 @@ type Count = number
 type NodeId = string
 
 type ValueDescriptor = {
-  int?: { total: Count, first: NodeId, example: number },
-  float?: { total: Count, first: NodeId, example: number },
-  date?: { total: Count, first: NodeId, example: string },
-  string?: { total: Count, first: NodeId, example: string, empty: Count },
-  boolean?: { total: Count, first: NodeId, example: boolean },
-  array?: { total: Count, first: NodeId, item: ValueDescriptor },
-  relatedNode?: { total: Count, first: NodeId, nodes: { [NodeId]: Count } },
-  relatedNodeList?: { total: Count, first: NodeId, nodes: { [NodeId]: Count } },
-  object?: { total: 0, first: NodeId, props: { [string]: ValueDescriptor } },
+  int?: TypeInfoNumber,
+  float?: TypeInfoNumber,
+  date?: TypeInfoDate,
+  string?: TypeInfoString,
+  boolean?: TypeInfoBoolean,
+  array?: TypeInfoArray,
+  relatedNode?: TypeInfoRelatedNode,
+  relatedNodeList?: TypeInfoRelatedNode,
+  object?: TypeInfoObject,
 }
+
+abstract type TypeInfo = {
+  first?: NodeId | void, // Set to undefined if "del"
+  total: Count,
+}
+
+type TypeInfoString = TypeInfo & {
+  empty: Count,
+  example: string,
+}
```

Comment:
Cool, I like it this way! Initially, this structure was relatively small and it was easier to read it inline but it kinda grew and it makes sense to extract those into separate "types".

Reply:
I wish it were easier to make this clear in the code. Like, create the object with these props when we actually need it, rather than a generic TypeInfo object that then gets enhanced (I mean in the code, not the type declaration). But with the current architecture that was a bridge too far.